sidecar_acceptor.go

```diff
@@ -0,0 +1,490 @@
+						"message: %v", err)
+				}
+
+			case *auctioneerrpc.ServerAuctionMessage_Finalize:
+				a.matchFinalize(msg.Finalize)
+
+			default:
+				sdcrLog.Debugf("Received msg %v from "+
+					"auctioneer on sidecar client: %v", msg)
+			}
+
+		case <-a.quit:
+			return
+		}
+	}
+}
+
+// Stop stops the sidecar acceptor.
+func (a *SidecarAcceptor) Stop() error {
+	var returnErr error
+	if err := a.client.Stop(); err != nil {
+		sdcrLog.Errorf("Error stopping auctioneer client: %v", err)
+		returnErr = err
+	}
+
+	a.acceptor.Stop()
+	close(a.quit)
+
+	a.wg.Wait()
+
+	return returnErr
+}
+
+// RegisterSidecar derives a new multisig key for a potential future channel
+// bought over a sidecar order and adds that to the offered ticket. If
+// successful, the updated ticket is added to the local database.
+func (a *SidecarAcceptor) RegisterSidecar(ctx context.Context,
+	ticket *sidecar.Ticket) error {
+
+	// The ticket needs to be in the correct state for us to register it.
+	if err := sidecar.VerifyOffer(ctx, ticket, a.signer); err != nil {
+		return fmt.Errorf("error verifying sidecar offer: %v", err)
+	}
+
+	// Do we already have a ticket with that ID?
+	_, err := a.store.Sidecar(ticket.ID, ticket.Offer.SignPubKey)
+	if err != clientdb.ErrNoSidecar {
+		return fmt.Errorf("ticket with ID %x already exists",
+			ticket.ID[:])
+	}
+
+	// First we'll need a new multisig key for the channel that will be
+	// opened through this sidecar order.
+	keyDesc, err := a.wallet.DeriveNextKey(
+		ctx, int32(keychain.KeyFamilyMultiSig),
+	)
+	if err != nil {
+		return fmt.Errorf("error deriving multisig key: %v", err)
+	}
+
+	ticket.State = sidecar.StateRegistered
+	ticket.Recipient = &sidecar.Recipient{
+		NodePubKey:       a.nodePubKey,
+		MultiSigPubKey:   keyDesc.PubKey,
+		MultiSigKeyIndex: keyDesc.Index,
+	}
+	if err := a.store.AddSidecar(ticket); err != nil {
+		return fmt.Errorf("error storing sidecar: %v", err)
+	}
+
+	return nil
+}
+
+// ExpectChannel informs the acceptor that a new bid order was submitted for the
+// given sidecar ticket. We subscribe to auction events using the multisig key
+// we gave out when we registered the ticket.
+func (a *SidecarAcceptor) ExpectChannel(ctx context.Context,
+	t *sidecar.Ticket) error {
+
+	a.Lock()
+	defer a.Unlock()
+
+	if t.Order == nil {
+		return fmt.Errorf("order in sidecar ticket is missing")
+	}
+
+	nonce := t.Order.BidNonce
+	_, ok := a.pendingSidecarOrders[nonce]
+	if ok {
+		return fmt.Errorf("sidecar with order nonce %x is already "+
+			"registered", nonce[:])
+	}
+
+	// We didn't know about this ticket for this nonce before so let's now
+	// update its state in the database and start expecting a channel for it
+	// now.
+	t.State = sidecar.StateExpectingChannel
+	if err := a.store.UpdateSidecar(t); err != nil {
+		return fmt.Errorf("error updating sidecar: %v", err)
+	}
+
+	a.pendingSidecarOrders[nonce] = t
+
+	return a.client.SubscribeAccountUpdates(ctx, &keychain.KeyDescriptor{
```

Comment:
I'm a bit confused here. Is `ExpectChannel` meant to be long running, or is it called and immediately returns?

Reply:
Yes, you are right, the name `SubscribeAccountUpdates` isn't conveying its purpose correctly. I renamed the method and also added a comment here.